File: December2022/Java/MaximumDifferenceBetweenNodeAndAncestor.java
// @saorav21994

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result = 0;
    public void find(TreeNode root, int min, int max) {
        if (root == null) {
            return;
        }
        int tmp = Math.max(Math.abs(root.val - max), Math.abs(root.val - min));
        result = Math.max(tmp, result);
        min = Math.min(root.val, min);
        max = Math.max(root.val, max);
        find(root.left, min, max);
        find(root.right, min, max);
        return;
    }
    
    public int maxAncestorDiff(TreeNode root) {
        result = 0;
        if (root == null) {
            return 0;
        }
        find(root, root.val, root.val);
        return result;
    }
}

File: December2022/C++/maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    
    pair<int,int> findMaxDiff(TreeNode* root, int &maxDiff){
        
        if(root->left==NULL && root->right == NULL){
            return {root->val, root->val};
        }
        
        int node = root->val;
        
         pair<int,int>leftMax = {node,node};
         pair<int,int>rightMax = {node,node};
        
        if(root->left){
            
            leftMax = findMaxDiff(root->left, maxDiff);
        }
        if(root->right){
            
            rightMax = findMaxDiff(root->right, maxDiff);
            
        }
        
        maxDiff = max({maxDiff,abs(leftMax.first-node),abs(leftMax.second-node),abs(rightMax.second-node),abs(rightMax.first-node)});
        
        
        return {max({node,leftMax.first,rightMax.first}), min({node,leftMax.second,rightMax.second})};
    }
public:
    int maxAncestorDiff(TreeNode* root) {
        
        int maxDiff = 0;
        
        findMaxDiff(root, maxDiff);
        
        return maxDiff;
    }
};